[2기-P 이연우 ] Spring Boot JPA 게시판 구현 코드 리뷰 부탁드립니다

ynoolee opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- Post : User 는 다대일 단방향 관계입니다.
- 현재 사용자 로그인 기능이 따로 없기 때문에, 게시글 작성 요청을 보내올 때마다 userId 도 함께 전달받도록 하였습니다. 서비스에서는 존재하는 사용자인지 여부를 확인한 후, 게시글 생성시 존재하지 않는 사용자이면 권한이 없다는 예외를 던지도록 하였습니다 ( 그런데 예외 처리 핸들링은 아직 하지 않았슴다 ) 
- BaseEntity → User 에도 createdBy 가 있는 것이 어색했고, createdBy 보다는 updatedAt 이 있는 것이 더 적절하다고 생각하여 이들을 BaseEntity로 작성하였습니다.
- post 관련 DTO 들은 PostDto 라는 하나의 클래스의 nested class 로 위치시켰습니다.
    - 관련된 DTO 들을 하나의 클래스에서 다룰 수 있다는 장점이 있다고 생각했고
    - 이렇게 함으로서, conversion 과 관련된 코드들 역시 하나의 클래스 내부에서 관리된다는 장점 또한 존재한다고 생각했습니다.
- 응답 객체는 세가지가 존재합니다 ( Success, BindingError, Error )Bean validation 에 대한 에러 메시지에 대한 응답구조는 따로 정의하였습니다

## 완성되지 않은부분 - Exception

완성..이란 것은 없을지도 모르지면 현재 눈에 뜨이게 부족한 부분들이 존재합니다. 일단 제출을 하고 추가하고자 하였습니다

- 커스텀 예외들에 대한 클래스만이 존재하고, 아직 이들에 대하여 핸들러를 통한 *예외처리 코드를 아직 작성하지 않았습*니다
    - 커스텀 exception 들에 대한 구현은 아직 아무것도 하지 않았습니다 .
    - 커스텀 예외들에 대한 처리를 전혀 하지않고 있습니다.
- 에러들을 관리하는 부분에 대해 아직 생각해 보지 않았습니다.
    - 에러메시지들도 부실하게 작성되어있슴다.

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
궁금한 점은 조금 더 정리해서 함께 얘기해보도록 하겠습니다!

ynoolee committed:
Init project

ynoolee committed:
chore: application.yaml 설정

ynoolee committed:
feat: JpaAuditing 활성화

ynoolee committed:
feat: BaseEntity 추가

ynoolee committed:
refactor: BaseEntity 를 추상클래스로 변경하고 필드의 접근제한자를 변경

ynoolee committed:
feat(domain,test): Board 추가 및 Board 생성 테스트

ynoolee committed:
feat(domain) : User 구현 및 Post 와의 연관관계 추가

- 필요한 Exception 추가
- 다대일 단방향 연관관계

ynoolee committed:
feat(test) : Post 테스트 추가

ynoolee committed:
feat : 취미 열거 타입 추가

ynoolee committed:
refactor: JpaConfig 분리

- MockMvc 테스트과정에서 발생하는 JPA metamodel must not be empty 에러 해결

ynoolee committed:
feat(service): 게시글 생성,수정,다건,단건 조회 기능 추가

ynoolee committed:
feat: UserDto 추가

ynoolee committed:
feat: UserRepository 추가

ynoolee committed:
feat: 권한실패예외 추가 및 생성실패 예외 리팩토링

- 게시글 작성권한이 없는 경우에 대한 예외 추가( 존재하지 않는 사용자인 경우 )
- 생성 실패 객체 타입을 출력하는 것으로 변경

ynoolee committed:
feat: 게시글 작성 권한이 존재하지 않는 경우에 대한 예외처리

- 레포지토리가 아닌 서비스에 의존하는 것으로 변경

ynoolee committed:
feat: 유저 서비스 추가

ynoolee committed:
feat : PostRestController 와 Api 성공 응답 객체 추가

ynoolee committed:
feat : 컨트롤러 아규먼트인 페이징 객체 역직렬화 테스트

ynoolee committed:
refactor : User - DTO 변환로직 추가로 인한 변경

ynoolee committed:
feat: ErrorResponse 와 BindingErrorResponse 추가

- 빈 검증과 같은 바인딩 과정에서 발생하는 예외에 대한 응답은 분리하였다 ( 추후 다른 예외들 또한 MessageSource 로 관리 할 수 있을 것 같음 )

ynoolee committed:
feat: 전역 예외 처리기 추가

ynoolee committed:
refactor: PostDto.Info 생성자 -> 정적 팩토리 메소드로 변경

ynoolee committed:
fix: 사용자 생성 시 이메일 세팅

ynoolee committed:
feat: 유저서비스 클래스에 readOnly transaction 어노테이트 및 메소드 단위 트랜잭션 오버라이드 추가

ynoolee committed:
refactor: url pattern 에 /api 를 추가

ynoolee committed:
refactor: Post 빌더에 id 필드 추가

ynoolee committed:
feat: 직렬화된 DTO 제약조건 검증 테스트 추가

ynoolee committed:
feat: 응답 객체 직렬화 테스트 추가

ynoolee committed:
refactor : 메소드 네임 변경

ynoolee committed:
refactor : 메소드 네임 변경 ( create -> store)

ynoolee committed:
feat : 컬럼에 대한 updatable = false 설정 추가

ynoolee committed:
feat : User 로부터 UserDto.CreateRequest 를 생성하는 정적 팩토리 메소드 추가

ynoolee committed:
feat : BoardRepository 테스트 추가

ynoolee committed:
refactor: 패키지 변경

ynoolee committed:
refactor: api 응답 객체 클래스들의 패키지 변경

ynoolee committed:
refactor: DTO 클래스들의 패키지를 변경

- 중첩 클래스를 사용하며 현재처럼 하나의 톱레벨 dto 클래스만 존재하는 경우에는 하위 패키지에 두는 것보다는 꺼내두는게 더 적절하다고 생각했다
- 만약 Request, Response DTO 를 나누게 된다면 하위 패키지를 만드는 것을 고려 해 볼 수 있을 듯 하다

ynoolee committed:
feat: UserRestController 새로운 유저 생성 기능 추가

ynoolee committed:
feat: Restdocs 추가

ynoolee committed:
refactor: PostDto 에 대한 static import 및 PostDto 중첩클래스들 네이밍 변경 -> Post 를 삽입

- static import 를 하면서 Info 와 같은 네이밍은 불명확하다고 생각하게 되어 Dto 중첩 클래스드들의 네이밍에 Post 를 붙여주었음

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/commons/domain/BaseEntity.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+@Setter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
```
`@EntityListeners(AuditingEntityListener.class)` 어노테이션은 어떤 동작을 하나요? 🤔

ynoolee replied:
EntityListener 의 파라미터로 콜백을 요청할 클래스를 지정하여 사용해요. 엔티티 리스너에서는 @PrePersist, @PreUpdate 어노테이티드된 메소드들을 갖고 있는데 이들을 통해 캡쳐해야하는 정보들을 감지할 수 있는 것으로 알고 있습니당. 정확한 시점은 언제다 라는 것은 저도 잘 모르지만, 엔티티를 디비에 저장하기 전(?)에 영속성과 관련된 리스너에서 이러한 콜백 메소드들을 사용하여 엔티티의 속성값을 디비에 자동으로 업데이트 하는 것이 가능해지는 것으로 알고 있습니다.

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
+
+@Configuration
+@EnableJpaAuditing
```
`@EnableJpaAuditing` 어노테이션은 어떤 동작을 하나요?  🤔

ynoolee replied:
이게 있어야 Auditing 기능이 활성화 될 수 있습니다!

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/post/Post.java:
```diff
@@ -0,0 +1,119 @@
+@Entity
+public class Post extends BaseEntity {
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
```
IDENTITY 전략을 사용한 이유는 무엇인가요?  🤔

ynoolee replied:
과제의 요구사항인 auto increment 를 위해서 사용했습니다!

zxcv9203 replied:
SEQUENCE 전략을 사용하지 않고 IDENTITY를 사용한 이유는 무엇인가요?

ynoolee replied:
H2 의 경우는 SEQUENCE 전략을 사용하면 auto_increment 와 같은 효과를 줄 수 있겠지만 MySQL 을 고려한다면 IDENTITY 를 사용하는 것이 좋을 것이라 생각했습니다 MySQL 에 대한 생각을 기반으로 했던 것 같습니다

zxcv9203 replied:
IDENTITY 전략 같은 경우 쓰기 지연이 불가능한데 이부분에 대해서는 어떻게 생각하시나요?

https://www.baeldung.com/hibernate-identifiers
https://docs.jboss.org/hibernate/orm/5.0/mappingGuide/en-US/html/ch06.html#identifiers-generators-identity

ynoolee replied:
맞습니다 쓰기 지연이 불가능하여 MySQL 에서도 IDENTITY 전략을 사용하여 Batch INSERT 에서 문제가 있는 것으로 알고 있습니다. 흠..이 사실을 알고 있음에도 저는 별 생각없이 Auto_increment 를 위해서는  IDENTITY 를 사용해왔던 것 같아 반성하게 됩니다. 
 일단 현재처럼 H2 를 사용할 경우에는 SEQUENCE 가 더 적절해 보입니다.  그리고 SEQUENCE 를 사용할 때에도 성능의 이점을 보려면, 매 번 next value 를 요청하지 않기 위한 적절한 설정이 필요하지 않은가? 라는 생각이 들었습니다!

그리고 MySQL 에서 auto_increment 를 사용하려 할 때 어떻게 할 것인지에 대해 팀원분들과 논의해보고 싶습니다 !!
개인적으로 만약 IDENTITY 를 사용하면서 bulk insert를 해야만 하는 상황이라면 해당 부분에 대해서는 JDBC 를 사용하는 방식이 있을 듯 합니다. ( 직접 해 본적은 없슴다.. )

zxcv9203 replied:
제가 연우님 말씀을 잘못이해한 걸 수도 있지만 기본적으로 SEQUENCE 전략의 경우 allocationSize가 기본적으로 50이 주어지기 때문에 기본값으로 둬도 문제 없을 것 같습니다.

그리고 mysql을 사용하는 상황에서 지연로딩을 사용하고 싶다면 Table 전략을 사용해야 할 것 같습니다.
단, 이방법은 Sequence에 비해 성능 이슈가 있어 많이 사용하지 않는다고 들었습니다

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/post/PostRestController.java:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	@GetMapping
+	public ResponseEntity<SuccessResponse<List<PostInfo>>> getPostsByPaging(Pageable pageable) {
```
ResponseEntity를 사용하시는 이유가 어떤건가요?  🤔

ynoolee replied:
헉..제가 이제까지 항상 ResponseEntity 를 사용해왔네요. 이것을 사용하면 @ResponseStatus 를 사용하지 않고도 statuscode 를 쉽게 설정해줄 수 있다고 무지성으로 사용해왔던 것 같습니다. 좀 더 알아봐야할 것 같습니다

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/post/PostService.java:
```diff
@@ -0,0 +1,110 @@
+				p.getTitle(),
+				p.getContent(),
+				p.getCreatedBy()))
+			.collect(Collectors.toList());
```
조회시에 해당 로직을 수행하는 것 같은데 따로 메서드로 분리하는 것은 어떠신가요?  🤔

ynoolee replied:
분리하는게 좋을 것 같습니다 수정하겠습니다!

zxcv9203 commented on src/main/java/com/prgrms/boardjpa/user/domain/User.java:
```diff
@@ -0,0 +1,165 @@
+	@Enumerated(EnumType.STRING)
+	@Column(name = "hobby")
+	private Hobby hobby;
+	// TODO : nullable = true 라면 getter 에서는 Optional 을 반환해줘야 할까? getter 가 Optional 을 반환하는 것은 바람직한가?
```
저는 null이 들어가도록 하는 것보다 NONE이라는 필드를 추가하여 없다는 것을 표시해서 무조건 값을 가지게 하는 방향으로 할 것 같습니다. 🤔

ynoolee replied:
null 만을 고려하고 있었네요.. 이 경우에는 NONE 이라는 값을 두는게 더 좋을 것 같습니다 감사합니다!

ICCHOI replied:
저도 nullable = false로 지정하고 없으면 empty string을 가지게 하는 것이 좋을 것 같네용

zxcv9203 commented on src/test/java/com/prgrms/boardjpa/post/BoardRepositoryTest.java:
```diff
@@ -0,0 +1,134 @@
+public class BoardRepositoryTest {
+
+	@Autowired
+	PostRepository postRepository;
```
접근제어자를 private으로 하지 않은 이유가 있으신가요?  🤔

ynoolee replied:
없습니다..깜빡했습니다 🥲

zxcv9203 reviewed (approved):
JPA 게시판 미션 고생하셨습니다 🎉
코드에서 다양한 방법을 적용하려고 하시는 것이 느껴졌습니다 👍👍👍

ICCHOI commented on src/main/java/com/prgrms/boardjpa/user/domain/Hobby.java:
```diff
@@ -0,0 +1,7 @@
+package com.prgrms.boardjpa.user.domain;
+
+public enum Hobby {
```
취미가 enum으로 관리되어야 하는 이유가 있나요?
특정 취미가 아닌 취미는 등록할 수 없는건가요?

ynoolee replied:
넵 현재는 그렇게 생각하면서 작성했었습니다!

ICCHOI commented on src/main/java/com/prgrms/boardjpa/user/UserRestController.java:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	@PostMapping
+	public ResponseEntity<SuccessResponse<UserDto.Info>> create(@RequestBody @Valid UserDto.CreateRequest createRequest) {
```
성공할 때 UserDto.Info를 반환하는 이유가 궁금합니다!

ynoolee replied:
사용자에서는 리소스의 생성을 요청한 것이기 때문에, 생성된 리소스에 대한 정보로 응답을 해 줘야 한다고 생각했습니다!

ICCHOI commented on src/main/java/com/prgrms/boardjpa/commons/api/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+package com.prgrms.boardjpa.commons.api;
+
+public class ErrorResponse {
```
ErrorResponse로 message만 받는다면 굳이 다른 클래스로 나눌 필요가 있을까 생각이 드는데 어떠신가요?

ynoolee replied:
응답에 대하여 성공에 대한 응답과 실패에 대한 응답을 분리할지 하나의 응답으로 처리할지에 대한 고민이 있었습니다
하나의 응답으로 실패상황과 정상적으로 처리되는 상황을 표현하기 위해서는 boolean success 나 에러코드를 통해서 알려줄 수 밖에 없다는 생각을 했습니다. 
- 현재는 별개의 에러코드를 정의해두지 않았고, 저는 응답데이터의 json 데이터로서 상태코드를 포함할 생각이 없었습니다

그리고 이 경우에는 정상적인 처리 결과 응답하는 데이터를 위한 필드와, 에러에 대한 정보를 나타내는 필드가 항상 같이 있어야 하는 것으로..밖에 생각이 안들었습니다. 이 경우, 정상적인 처리를 하는 상황에도 응답객체에는 항상 error 필드가 있는( 물론 어노테이션을 통해 null 인 경우는 Json 으로는 출력이 안되는 설정같은것이 가능하겠지만) 것이 마음에 들지 않는다..라는 생각이 들었습니당

따라서 이번에는 에러상황에 대한 응답과 성공적인 상황에 대한 응답을 분리하였습니다!

ICCHOI commented on src/main/java/com/prgrms/boardjpa/commons/api/BindingErrorResponse.java:
```diff
@@ -0,0 +1,50 @@
+import lombok.Getter;
+
+@Getter
+public class BindingErrorResponse {
```
BindingErrorResponse는 어떤 생각을 하시면서 구성했는지 궁금하네요!

ynoolee replied:
validator 의 검증 결과로 BindingResult 에 담겨진  fieldError 들에 대한 별개의 응답 형식을 만들고싶었습니다!
![image](https://user-images.githubusercontent.com/53856184/171228076-fd97ee9c-21d4-4a5f-be32-17aae1a24815.png)

ICCHOI reviewed (commented):
연우님 고생하셨습니다!
전체적으로 고민을 하면서 짜신 흔적이 많이 느껴지는 것 같아요
세 분 연속해서 코드를 봐서 그런가 연우님 리뷰가 제일 적네요.. 
혹시나 나중에 보고 추가할 내용이 있으면 추가하겠습니다. 고생하셨어요 :)

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/user/domain/User.java:
```diff
@@ -0,0 +1,165 @@
+		this.hobby = hobby;
+	}
+
+	private void validateName(String name) {
+		Assert.hasText(name, "닉네임은 2글자 이상이어야 합니다");
+
+		if (name.length() < MIN_NAME_LENGTH) {
+			throw new IllegalArgumentException("닉네임은 2글자 이상이어야 합니다");
+		}
+	}
```
@Size 어노테이션을 통해서 validation을 진행할 수 있다고 알고있는데, 혹시 따로 메소드로 확인하신 이유가 있으신가요?

ynoolee replied:
@Size 어노테이션을 사용해 보았는데 제가 테스트 했을 때는 해당 제약조건에 대한 검증이 제가 원하던 시점(객체 생성시점)에 이루어지는 것이 아님을 확인하였습니다 ! 
그래서 엔티티에 대한 여러 검증로직들을 별개로 작성하였는데, @Size 와 같은 어노테이션이 객체 생성시점에 적용이 되는 거였을까용..?

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/user/domain/User.java:
```diff
@@ -0,0 +1,165 @@
+
+	private void validateEmail(String email) {
+		Assert.hasText(email, "이메일은 비어있을 수 없습니다");
+	}
```
이 부분 또한 @NotEmpty 또는 @NotBlank 등을 이용해서 validation을 진행할 수 있다고 알고있는데, 혹시 메소드로 따로 구현하신 이유가 있으신가요?

ynoolee replied:
위와 같습니다!

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/user/UserRestController.java:
```diff
@@ -0,0 +1,31 @@
+
+	@PostMapping
+	public ResponseEntity<SuccessResponse<UserDto.Info>> create(@RequestBody @Valid UserDto.CreateRequest createRequest) {
+		return ResponseEntity.ok(
+			SuccessResponse.of(
+				userService.store(createRequest)
```
ResponseEntity를 사용하면 body 뿐만 아니라 header, status코드도 적용할 수 있군요!👍
이번엔 header는 없는것 같은데, 그러면 status 코드를 위해 사용하신거겠죠??😁

ynoolee replied:
넵 상태코드를 위해서 사용하기도 하였지만, 이렇게 해야지만 직렬화가 이루어진다고 생각해왔던 것 같습니다🥲
@RestController 클래스 내부의 메소드들은 자동적으로 @ResponseBody 어노테이션이 적용되어 리턴하는 객체에 대한 직렬화가 이루어진다는 것을 잊고 사용해오고 있었습니당..

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/post/Post.java:
```diff
@@ -0,0 +1,119 @@
+	@Column(name = "content", columnDefinition = "TEXT", nullable = false)
+	private String content;
+
+	@Column(name = "writer", nullable = false, updatable = false)
```
@Column에 updatable 이라는 설정도 있군요👍

ynoolee replied:
넵 이 설정에 따라 엔티티를 업데이트 하는 save 메소드 호출시에 날아가는 쿼리가 달라짐을 확인하여 따로 설정을 해 주었습니다. 저는 게시글의 경우 제목과 컨텐츠만을 편집할 수 있다고 가정하여 해당 필드들에 대해서만 updatable 에 대한 기본값(true) 를 사용하였습니다

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/post/Post.java:
```diff
@@ -0,0 +1,119 @@
+	@Column(name = "content", columnDefinition = "TEXT", nullable = false)
+	private String content;
+
+	@Column(name = "writer", nullable = false, updatable = false)
+	private String createdBy;
+
+	@ManyToOne
+	@JoinColumn(name = "user_id", nullable = false, updatable = false)
+	private User writer;
```
User writer는 db에 user_id로 저장이 되고 위의 String createdBy는 db에 writer로 저장이 되면 뭔가 헷갈리지 않을까요...? 제 개인적인 생각입니다! 다른 분들은 어떠신지 궁금하네요😊

ynoolee replied:
다대일 단방향 관계이긴 하지만, 보통 게시글의 경우 작성자의 이름이 항상 표시된다는 것을 생각하였을 때, 굳이 User 에 대한 정보도 같이 가져오는 것 보다는 userName 은 Post 에 대한 정보로서 추가적으로 갖고 있는게 좋지 않을까? 라는 생각을 하였습니다.
혹시 db table 의 column name 에서의 혼동이 오지 않을까? 라는 말씀이시라면 writerId, writer 로 컬럼 네임을 명시해주는게 좋겠다 라는 생각이 듭니다! 일단 이렇게라도 변경해보겠습니다!😀

ICCHOI replied:
연우님의 고민이 공감이 가는데, 이게 객체 지향적인 설계인지는 의문이 드네요. 
User와 Post의 관계가 명확하게 정의 되어 있는데 User의 특정 컬럼을 자주 호출 해야한다고 해서 그 정보를 추가적으로 가지고 있는다고 한다면, 객체지향적보단 데이터 중심적 개발이 아닌가 하는 생각이 듭니다!
만약 성능상의 이유로 그러한 설계를 고려한다면 Index를 사용하는게 맞는 방법일 것 같습니다. (https://velog.io/@ljinsk3/JPA%EB%A1%9C-%EC%9D%B8%EB%8D%B1%EC%8A%A4-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0)

JiyoungMa commented on src/main/java/com/prgrms/boardjpa/post/PostRepository.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface PostRepository extends JpaRepository<Post, Long> {
+	List<Post> findAllBy(Pageable pageable);
```
Page<Post> findAll(Pageable pageable) 라고 기본적으로 제공해주는 메소드를 사용하지 않고 List<Post>형태로 리턴하는 메소드를 만드신 이유가 있으신가요?
그리고 이건 궁금한 부분인데 이 부분을 따로 구현하지 않아도 작동하는 건가요?!?!

ynoolee replied:
1. -> 요구사항의 페이지 라는 것에 대해 저는 특정 페이지의 데이터들만을 가져오는 것으로 생각하였고 따라서 List 를 반환하도록 하는 것이 필요하다 생각하였습니다!
기본적으로 제공되는 Page 로 반환할 findAll(Pageable pageable) 의 경우에는 카운트 쿼리가 추가적으로 날아기 때문에 사용하지 않았습니다!
그런데 [이런글](https://www.inflearn.com/questions/32481)도 존재하네용 
2. -> PagingAndSortingRepository 에는 	Page<T> findAll(Pageable pageable) 만이 선언되어있고 SimpleJpaRepository 에서도 Pageable 을 받아 List 로 반환하는 것을 찾지 못해 따로 쿼리 메소드를 선언하였습니당

JiyoungMa commented on src/test/java/com/prgrms/boardjpa/post/PostRestControllerTest.java:
```diff
@@ -0,0 +1,323 @@
+@AutoConfigureMockMvc
+@AutoConfigureRestDocs(uriScheme = "https", uriHost = "docs.api.com")
+@WebMvcTest(PostRestController.class)
+public class PostRestControllerTest {
```
User는 ControllerTest를 하지 않으신 특별한 이유가 있으신가요?🤔

ynoolee replied:
없습니다! User 에 대해서는 동작에 관한 실행을 직접 해보기만 하고 제출하였습니다 
유저에 대해서도테스트 코드가 필요하다고 생각합니다🥲

JiyoungMa reviewed (commented):
연우님 수고하셨습니당~! Dto안에 record가 있는건 특이하네요! Dto pacakage안에 class로 구현하는 것과 비교해서 어떤 장점이 있을지 궁금합니다! 이 부분에 대한 다른 분들의 의견도 궁금하네요😊

ynoolee committed:
refactor: 와일드 카드 타입으로 변경

ynoolee committed:
refactor: Hobby 를 nullable 할 수 없게 변경

- Hobby 의 기본값으로 NONE 상수가 올 수 있도록 한다

ynoolee committed:
refactor: 기본키 생성전략을 AUTO 로 변경

ynoolee committed:
refactor: mysql 설정 추가

- AUTO 전략을 사용할 경우 현재 Hibernate 버전에서는 mysql 에서 TABLE 전략을 사용하는 것을 확인하기위해 mysql 설정을 추가하였다
- 테스트코드에서는 H2 설정을 사용하게 하였음

ynoolee committed:
refactor: 클래스네임변경(board->post) 및 필드 접근제한자 변경 (private-package -> private)

ynoolee committed:
refactor: Pageable -> SimplePage 로 변경

- page, size 요청 파라미터에 대한 옵션만을 제공하는 Page 로 변경

ynoolee committed:
refactor: 이전 커밋에서 빠진 설정 파일( 메인 application.yaml)

ynoolee committed:
refactor: CreationFailException 리팩토링

- cause 를 전달받는 생성자 추가
- cause 를 활용하는 로그 메시지 추가

ynoolee committed:
refactor: 컬럼네임 변경(writer->writerName, user_id -> writer_id)

SeokRae reviewed (commented):
고생하셨습니다. 연우님
같이 이야기 나눴던 내용 고민해보시고, 학습해보시고 리펙토링 해보시죠 !

1. 패키징 관리에 대한 부분 고민 해보기
2. Restful API에 대한 학습 해보기
3. 테스트를 위한 로직은 제거하기
4. dto -> entity or entity -> dto에 대한 책임 클래스 고민 해보기
5. 스트림을 통한 비즈니스 로직 구현해보기

NownS commented on src/main/java/com/prgrms/boardjpa/commons/api/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,47 @@
+
+	@ExceptionHandler(value = {MethodArgumentNotValidException.class})
+	public ResponseEntity<BindingErrorResponse> handleBindingException(BindException bindException, Locale locale) {
+		log.info("BindException {}", bindException.getMessage());
```
BindingException의 로그를 info 수준으로 유지하는 이유가 있을까요??

ynoolee replied:
해당 경우 검증 기준에 어긋난 입력이라는 것은 프로그램 내부의 문제가 아니기 때문에 error 로 하기에는 적절하지 않다는 생각을 하였습니다. 
그래서 info 와 debug 중에 어떤 것을 해야할지 고민이 되었는데, 결론을 내리지 못해 일단 info 로 작성하였습니다. 
하지만 INFO 는 어플리케이션에서 정상적으로 발생한 어떤 일들 ( Post가 생성되었다면 게시글이 생성되었다는 일 ) 을 알려주는 것에 더 적합하다는 생각이 드네용
지금은 debug  레벨이 조금 더 맞을 것 같은데 진욱님 생각은 어떠신가요?!

NownS commented on src/main/java/com/prgrms/boardjpa/commons/domain/BaseEntity.java:
```diff
@@ -0,0 +1,35 @@
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+	protected BaseEntity() {
```
기본 생성자를 넣어 주신 이유는 무엇인가요?

NownS commented on src/main/java/com/prgrms/boardjpa/post/PostRestController.java:
```diff
@@ -0,0 +1,81 @@
+import com.prgrms.boardjpa.commons.api.SuccessResponse;
+
+@RequestMapping("/api/posts")
+@RestController
```
어노테이션의 순서는 어떻게 정하는 편이신가요? 저는 역할을 담당하는 어노테이션을 더 상단에 두고, 그 밑에 해당 역할에서 필요한 기능들을 배치하는 편인데, 저랑 반대로 배치해 두신 것 같아서 궁금해서 질문드려봅니다!

NownS reviewed (approved):
리뷰가 늦어서 죄송합니다ㅠㅠ
앞서 하신 분들이 많이 체크해주신 것 같아, 궁금한 점 위주로 포인트를 잡아보려고 했는데요, 코드를 쭉 보면서 이해하려고 하다 보니 가독성이 되게 좋고, 의도가 눈에 보이는 편이어서 질문거리가 그다지 많지는 않았네요.
고생하셨습니다!

ynoolee committed:
refactor: 패키지 변경

ynoolee committed:
refactor: Dto 의 컨버전 로직 -> PostConverter 사용으로 변경

- 서비스는 Converter 를 주입받아 Dto - Entity 간의 컨버전을 수행할 수 있다

ynoolee committed:
refactor: PostConverter 사용하는 것으로 리팩토링

ynoolee committed:
comment: @Transactional self-invocation 에 대한 주석 추가

ynoolee committed:
refactor(PostService,UserService) : 스트림을 사용해 리팩토링

ynoolee committed:
refactor: 변환로직을 UserConverter 로 분리

ynoolee committed:
feat: 게시글 좋아요 테스트 추가

ynoolee committed:
feat: 게시글 좋아요 기능 추가

ynoolee committed:
feat: 게시글 생성 응답에 대한 Location header 추가

ynoolee committed:
refactor: 메소드 네이밍 변경 ( pushLike -> like )

ynoolee committed:
refactor: 게시글 관련 서비스에 대한 로그 메시지 추가

- 일단 성공 상황에 대한 로그 추가

ynoolee committed:
feat: 유저가 좋아요 한 게시글 목록 불러오기 기능 테스트 추가

ynoolee committed:
feat: 유저가 좋아요 한 게시글 목록 불러오기 기능 구현 추가

ynoolee committed:
refactor: PostRepository 테스트코드 정리

ynoolee committed:
refactor: Post 엔티티에 롬복 사용

ynoolee committed:
refactor: like 메소드 네이밍 변경 (like -> likeBy, likedBy -> likeOf )

ynoolee committed:
feat: Post 도메인에 특정 사용자가 좋아요 했었는지 여부 알려주기 기능 추가

ynoolee committed:
refactor: 패키지 변경

ynoolee committed:
refactor: PostLikeServiceTest 클래스 패키지 변
